File: consistency-store/src/main/java/de/tudarmstadt/consistency/store/impl/ReadWriteRef.java
package de.tudarmstadt.consistency.store.impl;

import de.tudarmstadt.consistency.store.Operation;
import de.tudarmstadt.consistency.store.Ref;
import de.tudarmstadt.consistency.store.cassandra.CassandraRef;

/**
 * Created on 18.06.18.
 *
 * @author Mirko Köhler
 */
public abstract class ReadWriteRef<T> implements IReadWriteRef<T, ReadWriteRef<T>> {

	public void write(T value) throws Exception {
		handle(eventWrite(), value);
	}

	public T read() throws Exception {
		return handle(eventRead(), null);
	}

	protected abstract T handleRead() throws Exception;
	protected abstract void handleWrite(T value) throws Exception;

	@Override
	public <A, B> B handle(Operation<T, ReadWriteRef<T>, A, B> e, A param) throws Exception {
		return e.compute(this, param);
	}

	//TODO: Is it possible to make these static?
	private Operation<T, ReadWriteRef<T>, Void, T> eventRead() {
		return (ref, additionalParameter) -> ref.handleRead();
	}

	private Operation<T, ReadWriteRef<T>, T, Void> eventWrite() {
		return (ref, additionalParameter) -> {
			ref.handleWrite(additionalParameter);
			return null;
		};
	}







}

File: consistency-store/src/main/scala/de/tudarmstadt/consistency/store/scala/impl/ReadWriteRef.scala
package de.tudarmstadt.consistency.store.scala.impl

import de.tudarmstadt.consistency.store.scala.Ref

trait ReadWriteRef[Key, T] extends Ref[T, ReadWriteOp] {

		def read() : Option[T] =
			handle(ReadOp[T](), ())

		def write(value : T) : Unit =
			handle(WriteOp[T](), value)

		def :=(value : T) : T = {
			write(value)
			value
		}

		def apply() : T =	read() match {
			case Some(value) => value
			case None => throw new IllegalArgumentException(s"no value for reference available. reference: $this" )
		}


		override def handle[Parameter, Arity](op: ReadWriteOp[Parameter, Arity], parameter: Parameter): Arity =	op match {
			case ReadOp() => handleRead().asInstanceOf[Arity]
			case WriteOp() => handleWrite(parameter.asInstanceOf[T]).asInstanceOf[Arity]
		}


		protected def handleRead() : Option[T]
		protected def handleWrite(value : T) : Unit



	}